aips/aip-x-safe-auth-key-mapping.md

```diff
@@ -0,0 +1,221 @@
+the [Ledger key rotation docs].
+
+There are currently several issues with the `OriginatingAddress` table (which is
+supposed to be a one-to-one lookup table) that render the mapping unsafe in
+practice:
+
+1. Per [`aptos-core` #13517], `rotate_authentication_key_call` does not update
+   the `OriginatingAddress` table for an "unproven" key rotation without a
+   `RotationProofChallenge` (resolved in this AIP's reference implementation
+   with a new `set_originating_address` private entry function).
+1. During an operation that attempts to map an authentication key (which has
+   already been mapped) to a different originating address, the inner function
+   `update_auth_key_and_originating_address_table` overwrites the initial
+   mapping, rather than aborting. This oversight can lead to account loss if
+   someone accidentally attempts to rotate to the same authentication key twice
+   (resolved in this PR with `ENEW_AUTH_KEY_ALREADY_MAPPED` check), because they
+   will not be able to identify their account from private key alone unless they
+   keep an external record of the rotated accounts the private key in question
+   has been used to secure.
+1. Standard accounts that have not yet had their key rotated are not registered
+   in the `OriginatingAddress` table, such that two accounts can be
+   authenticated by the same authentication key: the original account whose
+   address is its authentication key, and another account that has had its
+   authentication key rotated to the authentication key of the original account.
+   Since `OriginatingAddress` is one-to-one, a dual-account situation can
+   inhibit indexing and OpSec (resolved in this PR with
+   `set_originating_address` private entry function).
+
+# Impact
+
+1. Without the changes proposed in this AIP's reference implementation,
+   unproven authentications (specifically those relying on
+   `rotate_authentication_key_call`) will result in an unidentifiable mapping,
+   such that users will be unable to identify accounts secured by their private
+   key unless they have maintained their own offchain mapping. This applies to
+   exotic wallets like passkeys.
+1. The overwrite behavior (described above) for
+   `update_auth_key_and_originating_address_table` can similarly result in an
+   inability to identify an account based on the private key.
+1. A user who authenticates two accounts with the same private key per the above
+   schema will experience undefined behavior during indexing and OpSec due to
+   the original one-to-one mapping assumption.
+
+# Summary
+
+The onchain key rotation address mapping has functional issues which inhibit
+safe mapping of authentication key to originating address for standard accounts.
+I propose resolving these issues with the reference implementation.
+
+# Proposed solution
+
+Assorted checks and extra function logic in [`aptos-core` #14309]
+
+# Alternative solutions
+
+Per @davidiw in a separate chat:
```

Comment:
For an AIP, we would need to be more precise about what the alternative solutions are.

I would remove the informal quote here and just precisely describe the alternative solution(s) considered.

Reply:
Addressed in 4e5fb6a